### PR TITLE
Run4-hgx250 Backport the corrections to HGCal geometry most relevant for partial wafers from 3 PR's #30323 #30413 #30425

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Make a number of corrections to HGCal geometry by back porting #30323 #30413 #30425. The corrections are more relevant for partial wafers and is applicable to all HGCal geometries.

#### PR validation:

Tested using runTheMatrix.py. But for using V11 geometry one needs the backport of Jean-Baptiste's code

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Backporting 3 PR's #30323, #30413 and #30425 which are integrated to master

### CODE REVIEW

**Comment by kpedro88** [comment]
@bsunanda can you make the PR title clearer?

**Comment by bsunanda** [comment]
@kpedro88 @silviodonato I think Jean-Baptiste's corrections need back porting before try out all the changes

File: Geometry/HGCalGeometry/src/HGCalGeometry.cc [hunk 1]
```diff
@@ -226,6 +226,26 @@ GlobalPoint HGCalGeometry::getPosition(const DetId& detid) const {
...
+  if (cellIndex < maxSize) {
+    const HepGeom::Point3D<float> lcoord(0, 0, 0);
+    if (mode_ == HGCalGeometryMode::Trapezoid) {
+      glob = m_cellVec2[cellIndex].getPosition(lcoord);
+    } else {
+      glob = m_cellVec[cellIndex].getPosition(lcoord);
+    }
+#ifdef EDM_ML_DEBUG
+    edm::LogVerbatim("HGCalGeom") << "getPositionTrap:: ID " << std::hex << detid << std::dec << " index "
+                                  << cellIndex < < < <
```
Comment (slava77) [inline Geometry/HGCalGeometry/src/HGCalGeometry.cc:242]: see a fix in #30513

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by silviodonato** [comment]
We are getting
```
----- Begin Fatal Exception 03-Jul-2020 13:28:37 CEST-----------------------
An exception of category 'BadGeometry' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 10 stream: 0
   [1] Running path 'FEVTDEBUGHLToutput_step'
   [2] Prefetching for module PoolOutputModule/'FEVTDEBUGHLToutput'
   [3] Prefetching for module HGCalBackendLayer1Producer/'hgcalBackEndLayer1Producer'
   [4] Prefetching for module HGCalConcentratorProducer/'hgcalConcentratorProducer'
   [5] Calling method for module HGCalVFEProducer/'hgcalVFEProducer'
Exception Message:
 EE:HSil= 0:1 type= 2 z= 1 layer= 7 wafer(u,v:x,y)= (-13,-9:17,-18) triggerCell(u,v:x,y)= (1,0:-20,-4)HGCalTriggerGeometry: Wafer (-13,-9) is not mapped to any trigger module. The module mapping should be modified. 
----- End Fatal Exception -------------------------------------------------
```
cc: @rekovic

**Comment by silviodonato** [comment]
please test with #30516

**Comment by kpedro88** [comment]
@silviodonato you may also need https://github.com/cms-sw/cmsdist/pull/6024 if it's not in the IB used for tests

**Comment by silviodonato** [comment]
please test with #30516, cms-sw/cmsdist#6024

**Comment by silviodonato** [comment]
please test with #30516, cms-sw/cmsdist#6024, #30531,#30523

**Comment by silviodonato** [comment]
please test with #30537

**Comment by silviodonato** [comment]
merge
the error in 23234.1001 step1 is fixed by https://github.com/cms-sw/cmssw/pull/30537

**Comment by cvuosalo** [comment]
+1

File: Geometry/HGCalCommonData/src/HGCalDDDConstants.cc [hunk 2]
```diff
@@ -1138,34 +1218,140 @@ std::pair<double, double> HGCalDDDConstants::waferPosition(
...
+        }
+      }
+    }
+  }
+#ifdef EDM_ML_DEBUG
+  if (debug)
+    edm::LogVerbatim("HGCalGeom") << "waferTypeRotation: Layer " << layer << " Wafer " << waferU << ":" << waferV
+                                  << " Index " << std::hex << wl << std::dec << ":"
+                                  << (itr != hgpar_->waferTypes_.end()) << " Type " << type << " Rotation " << rotn;
+#endif
```
Comment (slava77) [inline Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:1351]: this broke compilation of static analysis for this package.
Please fix.
Naively, this requires just moving the ifdef in the previous scope (before the ```}``` on L1345

Comment (makortel) [reply Geometry/HGCalCommonData/src/HGCalDDDConstants.cc:1351]: Or how about just using `LogTrace`. In the big picture this pattern of hiding non-`LogDebug`-level messages with `#ifdef EDM_ML_DEBUG` is confusing.

**Pull request merged**